pytorch_lightning/trainer/connectors/data_connector.py

```diff
@@ -117,19 +117,23 @@ def attach_dataloaders(
         # functions to overwrite with these implementations
         if train_dataloaders is not None:
             self.trainer.train_dataloader = None
-            model.train_dataloader = _PatchDataLoader(train_dataloaders)
+            train_dataloader = _PatchDataLoader(train_dataloaders, "train")
```

Comment:
@justusschock do you think it is possible to avoid patching all together? There are good reasons not to do it #6030 from a user's perspective. While I believe your PR solves the major issue, there could still be problems when the user wants to call their dataloader method to produce a fresh dataloader, for example in a callback (nothing is stopping them from doing so).

Some work has started here #7522

Reply:
@awaelchli @justusschock I think we need a centralization point for where dataloaders come from. Patching the dataloader methods onto the model is using the model as the source of truth, but the side effects are visible to the end user. Rather, we could create an internal DataHolder that could be used to pool the object that has the DataHooks available. This would also codify the priority/precedence across datamodules, lightning module, and dataloaders passed directly to the trainer.

It will also raise the question: what happens when both the datamodule and the lightning module have these hooks implemented? https://github.com/PyTorchLightning/pytorch-lightning/blob/e0605472306d6b95bf2616ab88f8c29f4498402e/pytorch_lightning/core/hooks.py#L455-L807

do we raise an exception? do we run only the datamodule's, if available? do we run both and if so, in what order?

Reply:
@ananthsub @awaelchli  let's continue this discussion in #6030 and merge this as a fix :)

I copied both your questions/statements there :)